src/Aspire.Hosting/ProjectResourceBuilderExtensions.cs

```diff
@@ -491,4 +491,48 @@ private static void SetAspNetCoreUrls(this IResourceBuilder<ProjectResource> bui
             }
         });
     }
+
+    private static void SetBothPortsEnvVariables(this IResourceBuilder<ProjectResource> builder)
+    {
+        builder.WithEnvironment(context =>
+        {
+            builder.SetOnePortsEnvVariable(context, "HTTP_PORTS", "http");
+            builder.SetOnePortsEnvVariable(context, "HTTPS_PORTS", "https");
+        });
+    }
+
+    private static void SetOnePortsEnvVariable(this IResourceBuilder<ProjectResource> builder, EnvironmentCallbackContext context, string portEnvVariable, string scheme)
+    {
+        if (context.EnvironmentVariables.ContainsKey(portEnvVariable))
+        {
+            // If the user has already set that variable, we don't want to override it.
+            return;
+        }
+
+        var ports = new StringBuilder();
```

Comment:
I started out using a `ReferenceExpressionBuilder`, but found that it wasn't working in Publish mode, because it was never getting evaluated. Is there any drawback to using strings directly here, since there is no port allocation going on in publish mode (iiuc)?

Reply:
This should use `ReferenceExpressionBuilder`, we don't want ports in the manifest in the env. These should point to bindings endpoint references, not concrete ports.

Reply:
Done.

Reply:
This causes `AspNetCoreUrlsNotInjectedInPublishMode` to fail, because it calls `EnvironmentVariableEvaluator.GetEnvironmentVariablesAsync()`, which expects endpoints to be allocated:

```
System.InvalidOperationException : The endpoint `http` is not allocated for the resource `projectName`.

EndpointReference.get_AllocatedEndpoint() line 105
EndpointReferenceExpression.ComputeTargetPort() line 194
EndpointReferenceExpression.GetValueAsync(CancellationToken cancellationToken) line 179
ReferenceExpression.GetValueAsync(CancellationToken cancellationToken) line 66
EnvironmentVariableEvaluator.GetEnvironmentVariablesAsync(IResource resource, DistributedApplicationOperation applicationOperation) line 28
ProjectResourceTests.AspNetCoreUrlsNotInjectedInPublishMode() line 276
```

But allocation only happens as part of `ApplicationExecutor.CreateContainersAndExecutablesAsync()`, which is not supposed to happen at Publish time.

One possible take is that this is not a real problem, but that it is wrong for the Publish test to try to call `EnvironmentVariableEvaluator.GetEnvironmentVariablesAsync(resource);`. It should instead generated the manifest and check off of that.

WDYT?

Reply:
You shouldn't need to look at the target port value so it should work now right?

Reply:
See the stack. The call to `EnvironmentVariableEvaluator.GetEnvironmentVariablesAsync` results in looking up the `AllocatedEndpoint`, which blows up.

Reply:
To be clear, this is not caused by the logic I had that was looking up TargetPort (and is now gone). It's specific to this one test.

Reply:
Fixed by calling `GetEnvironmentVariablesAsync` in Publish mode in the test.